tests/parsenum/main.c

```diff
@@ -0,0 +1,155 @@
+	if ((fabs(x - y) < 1) && (fabs(x - y) <= fabs(x + y) * 1e-6))
+		return (1);
+
+	/* Not equal. */
+	return (0);
+}
+
+#define TEST4_DO(str, var, min, max, target)			\
+	var parsenum_x;						\
+	int parsenum_ret;					\
+	fprintf(stderr,						\
+	    "Parsing \"%s\" as %s between %s and %s (incl.) yields "	\
+	    target "... ", str, #var, #min, #max);		\
+	parsenum_ret = PARSENUM(&parsenum_x, str, min, max);
+
+#define TEST2_DO(str, var, target)				\
+	var parsenum_x;						\
+	int parsenum_ret;					\
+	fprintf(stderr,						\
+	    "Parsing \"%s\" as %s yields " target "... ", str, #var);	\
+	parsenum_ret = PARSENUM(&parsenum_x, str);
+
+#define CHECK_SUCCESS(target)					\
+	if (parsenum_ret == 0) {				\
+		if (parsenum_equal((double)parsenum_x, (double)target)) { \
+			fprintf(stderr, "PASSED!\n");		\
+		} else {					\
+			fprintf(stderr, "FAILED!\n");		\
+			goto err0;				\
+		}						\
+	} else {						\
+		fprintf(stderr, "FAILED!\n");			\
+		goto err0;					\
+	}
+
+#define CHECK_FAILURE(target)					\
+	if (parsenum_ret != 0) {				\
+		if (errno == target) {				\
+			fprintf(stderr, "PASSED!\n");		\
+		} else {					\
+			fprintf(stderr, "FAILED!\n");		\
+			goto err0;				\
+		}						\
+	} else {						\
+		fprintf(stderr, "FAILED!\n");			\
+		goto err0;					\
+	}
+
+#define TEST4_SUCCESS(str, var, min, max, target) do {		\
+	TEST4_DO(str, var, min, max, #target);			\
+	CHECK_SUCCESS(target);					\
+} while (0)
+
+#define TEST4_FAILURE(str, var, min, max, target) do {		\
+	TEST4_DO(str, var, min, max, #target);			\
+	CHECK_FAILURE(target);					\
+} while (0)
+
+#define TEST2_SUCCESS(str, var, target) do {			\
+	TEST2_DO(str, var, #target);				\
+	CHECK_SUCCESS(target);					\
+} while (0)
+
+#define TEST2_FAILURE(str, var, target) do {			\
+	TEST2_DO(str, var, #target);				\
+	CHECK_FAILURE(target);					\
+} while (0)
+
+int
+main(int argc, char * argv[])
+{
+	(void)argc; /* UNUSED */
+	(void)argv; /* UNUSED */
+
+	WARNP_INIT;
+
+	_Pragma("GCC diagnostic ignored \"-Wtype-limits\"");
+
+	TEST4_SUCCESS("123.456", double, 0, 1000, 123.456);
+	TEST4_SUCCESS("nAn", double, 0, 0, (double)NAN);
+	TEST4_SUCCESS("inf", double, 0, (double)INFINITY, (double)INFINITY);
+	TEST4_SUCCESS("-InFiNitY", double, -(double)INFINITY, 0, -(double)INFINITY);
+	TEST4_SUCCESS("0", double, 0, 0, 0);
+	TEST4_SUCCESS("-0", double, 0, 0, -0.0);
+	TEST4_FAILURE("7f", double, 0, 1000, EINVAL);
+	TEST4_SUCCESS("0x7f", double, 0, 1000, 127);
+	TEST4_SUCCESS("-0x7f", double, -1000, 0, -127);
+	TEST4_SUCCESS("-1e2", double, -1000, 0, -100);
+	TEST4_SUCCESS("1e-2", double, 0, 1000, 0.01);
+
+	TEST4_SUCCESS("1234", size_t, -123, 4000, 1234);
+	TEST4_FAILURE("7f", size_t, 0, 1000, EINVAL);
+	TEST4_SUCCESS("0x7f", size_t, 0, 1000, 127);
+	TEST4_SUCCESS("0x77", size_t, 0, 1000, 119);
+	TEST4_SUCCESS("077", size_t, 0, 1000, 63);
+
+	TEST4_SUCCESS("0xffffffff", uint32_t, 0, (uintmax_t)0xffffffff,
+	    UINT32_MAX);
+	TEST4_FAILURE("0xfffffffff", uint32_t, 0, (uintmax_t)0xffffffff,
```

Comment:
It would be good to test the "value is within specified range, but outside of the maximum range of the unsigned type -> ERANGE" case.  So 

`TEST4_FAILURE("0xfffffffff", uint32_t, 0, (uintmax_t)0xfffffffff, ERANGE);`

for example.  (I thought this might have been what you were trying to do with the duplicate line.)

Also, we should probably check capitalization of hex values?